src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -125,102 +123,198 @@ private static List<LoginMethod> getParsedLoginMethods(Main main, AppIdentifier
+                        errors.add("No two loginMethods can have isPrimary as true.");
+                    }
+                    hasPrimaryLoginMethod = true;
+                }
+            }
+        }
+    }
+
+    private static String validateAndNormaliseRecipeId(String recipeId, List<String> errors) {
+        if (recipeId == null) {
+            return null;
+        }
+
+        // We don't perform any normalisation on the recipeId after reading it from request header.
+        // We will validate it as is.
+        if (!Arrays.asList("emailpassword", "thirdparty", "passwordless").contains(recipeId)) {
+            errors.add("Invalid recipeId for loginMethod. Pass one of emailpassword, thirdparty or, passwordless!");
+        }
+        return recipeId;
+    }
+
+    private static String validateAndNormaliseTenantId(Main main, AppIdentifier appIdentifier, String tenantId, String recipeId, List<String> errors)
             throws StorageQueryException, TenantOrAppNotFoundException {
         if (tenantId == null || tenantId.equals(TenantIdentifier.DEFAULT_TENANT_ID)) {
-            return;
+            return tenantId;
         }
 
         if (Arrays.stream(FeatureFlag.getInstance(main, appIdentifier).getEnabledFeatures())
                 .noneMatch(t -> t == EE_FEATURES.MULTI_TENANCY)) {
             errors.add("Multitenancy must be enabled before importing users to a different tenant.");
-            return;
+            return tenantId;
         }
  
+        // We make the tenantId lowercase while parsing from the request in WebserverAPI.java
+        String normalisedTenantId = tenantId.trim().toLowerCase();
         TenantConfig[] allTenantConfigs = Multitenancy.getAllTenantsForApp(appIdentifier, main);
         ArrayList<String> validTenantIds = new ArrayList<>();
         Arrays.stream(allTenantConfigs)
                 .forEach(tenantConfig -> validTenantIds.add(tenantConfig.tenantIdentifier.getTenantId()));
 
-        if (!validTenantIds.contains(tenantId)) {
+        if (!validTenantIds.contains(normalisedTenantId)) {
             errors.add("Invalid tenantId: " + tenantId + " for " + recipeId + " recipe.");
         }
+        return normalisedTenantId;
     }
 
-    private static void validatePasswordHashingAlgorithm(Main main, AppIdentifier appIdentifier, EmailPasswordLoginMethod emailPasswordLoginMethod, List<String> errors) throws TenantOrAppNotFoundException {
-        if(emailPasswordLoginMethod.hashingAlgorithm == null || emailPasswordLoginMethod.passwordHash == null) {
-            return;
-        }
+    private static Boolean validateAndNormaliseIsVerified(Boolean isPrimary) {
+        // No normalisation needs to be done for isVerified
+        return isPrimary;
+    }
+
+    private static long validateAndNormaliseTimeJoined(Number timeJoined) {
+        // We default timeJoined to 0 if it is null
+        return timeJoined != null ? timeJoined.longValue() : 0;
```

Comment:
need to check for timeJoined not being in the future? Or not being < 0?

Reply:
Fixed.